provider/alibabacloud/alibaba_cloud.go

```diff
@@ -652,31 +661,54 @@ func (p *AlibabaCloudProvider) updateRecords(recordMap map[string][]alidns.Recor
 				}
 			}
 			if !found {
-				p.createRecord(endpoint, target)
+				p.createRecord(endpoint, target, zones)
 			}
 		}
 	}
 	return nil
 }
 
-func (p *AlibabaCloudProvider) splitDNSName(fullName string) (rr string, domain string) {
-	name := strings.TrimSuffix(fullName, ".")
-	parts := strings.Split(name, ".")
-	if len(parts) < 2 {
-		rr = name
-		domain = ""
-	} else {
-		domain = parts[len(parts)-2] + "." + parts[len(parts)-1]
-		rrIndex := strings.Index(name, domain)
-		if rrIndex < 1 {
-			rrIndex = 1
+func (p *AlibabaCloudProvider) splitDNSName(dnsName string, zones []string) (rr string, domain string) {
+	name := strings.TrimSuffix(dnsName, ".")
+
+	// sort zones by dot, make sure subdomain at first
+	sort.Slice(zones, func(i, j int) bool {
+		return strings.Count(zones[i], ".") > strings.Count(zones[j], ".")
+	})
+
+	found := false
+
+	for _, filter := range zones {
+		if strings.HasSuffix(name, "."+filter) {
+			rr = name[0 : len(name)-len(filter)-1]
+			domain = filter
+			found = true
+			break
+		} else if name == filter {
+			domain = filter
+			rr = ""
+			found = true
+		}
+	}
+
+	if !found {
+		parts := strings.Split(name, ".")
+		if len(parts) < 2 {
```

Comment:
Why the magic number `2`?

Reply:
it was there from the beginning. it signifies when the domain record has just two parts -> `container-service.top`, then when alicloud sends a request to create the record it can use the resource record (RR). Should i create a constant instead?

Reply:
The purpose of this PR is to fix the incorrect assumption that domain administrative boundaries are at the second level. So while the logic did exist previously, I question whether it should continue after this PR. Since the function is now being passed the complete set of hosted zone domains, is there really a need to handle the `!found` case by applying the 2-level-domain heuristic?

Reply:
yes it makes sense. the records we want to retrieve/create are going to either equal to the hosted zone or resource record will have the hosted zone as suffix.

Reply:
the only problem i see is with this [test](https://github.com/allurisravanth/external-dns/blob/master/provider/alibabacloud/alibaba_cloud_test.go#L264)

this line `endpoints, err := p.Records(context.Background())` will will try to get the `hostedZoneDomains`  by making the call `hostedZoneDomains, err := p.getDomainList()` and since this is a test, the hostedZoneDomains value will be an empty string slice. 

that means a call to the splitDNSName function will result in an empty domain and this [condition](https://github.com/allurisravanth/external-dns/blob/master/provider/alibabacloud/alibaba_cloud_test.go#L254) fails

Reply:
That sounds like an invalid test. The test setup should ensure the mock returns a useful domain list.

Reply:
The test `TestAlibabaCloudProvider_Records` tests `p.Records` method which inturn calls the `splitDNSName` method. without the `if !found` clause the p.Records method will not receive any domains. 

`hostedZoneDomains` will always be an empty slice when we're using a mock alibabacloud provider.

Reply:
we can add a `if len(hostedZoneDomains) > 0` check to control how the domainName value is set

something like this

```
for _, domainName := range p.domainFilter.Filters {
    if len(hostedZoneDomains) > 0 {
		_, domainName = p.splitDNSName(domainName, hostedZoneDomains)
    }
...
}
```
this check needs to be added in multiple places

or we leave the `!found` condition in the `splitDNSName` method

please suggest.

Reply:
You need to fix the mock alibabacloud so that it returns a useful domain list from a `DescribeDomains` call.

Reply:
made the changes, domain list is now returned in the mocked `DescribeDomains` call